keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,859 @@
+
+Changing defaults is a strict no-go.
+
+### Risks and Mitigations
+
+Accidentally setting a timeout too low could DOS kubelet if many are used.
+Thus will mitigate by preventing timeout values that are too small or to often:
+exec(500ms minimum periodSeconds until success, then drop back to the original second range/value)
+grpc(200ms minimum until success, then drop back to the original second range)
+http(200ms minimum until success, then drop back to the original second range)
+
+note for exec, grpc, http .. for startup and readiness we will support the millisecond range sum to be greater than zero..
+in other words if they know 1.4 seconds is the right range and 1s just will not work they can ask for 1second plus 400mills
+vs having to go pick one of 1second where the first is known to fail or 2 where they know it's 600mills slower than it should
+be because 99% of runs will be less than 1.4s.
+
+Keep initialDelayMilliseconds where initialDelay is supported because zero is the current minimum initialDelay in seconds..
+thus no mitigation is necessary.
+
+The periodMilliseconds option for periodSeconds is mitigated to only being valid until success and with appropriate minimums
+specified above
+
+timeoutMilliseconds will not be implemented in alpha.. may investigate for possible inclusion in a subsequent KEP.
+
+Benchmarking currently WIP for alpha, will be implemented for beta to determine the ideal floor values. Several benchmarking tools are being considered (options include ones developed by Knative, Amazon, Red Hat). Benchmarking tooling will also work for similar areas (such as the PLEG work.)
+
+As @aojea notes sig-scalability is already measuring pod latency on startup,
+http://perf-dash.k8s.io/#/?jobname=gce-100Nodes-master&metriccategoryname=E2E&metricname=LoadHighThroughputPodStartup&Metric=pod_startup and adding a new metric
+to https://github.com/kubernetes/perf-tests may suffice.
+
+Question raised of how probes should be charged: charging is an issue for all exec/attach/port forward requests, and as usual the process in the container should be charged to the container/pod. This KEP isn't looking to make any major architectural changes in this regard.
+
+
+#### Overriding an existing field
+
+However, this change would be done on a probe by probe basis and would be opt-in. Meaning that the default behavior will not change unless a user specifically added the millisecond fields to their spec. Absent that specific opt-in by using the new fields, existing users not being impacted by this change (gate on/off) will be tested.
+
+## Design Details
+
+Draft PR:
+https://github.com/kubernetes/kubernetes/pull/107958/commits
+
+A quick highlighting of key points:
+
+* Implementing a feature gate (`SubSecondProbes`) to gate the feature
+
+* Adding three additional, optional fields to the Probe struct type:
+
+```go
+PeriodMilliseconds *int32 //note these are signed offset
+InitialDelayMilliseconds *int32
+TimeoutMilliseconds *int32
```

Comment:
I thought timeout was out of scope for the alpha?

Reply:
nod.. this was left over.. to describe how it could be implemented... in a follow up.